### ISSUE TITLE
Logprobs affected by temperature

### ISSUE THREAD

**Issue by dpaleka** [standardcompletions/rfcs#11]
**The interaction of `temperature` and `logprobs` / `top_logprobs` parameters produces different behavior on different providers.**

There are three reasonable default behaviors:
- (1) Logprobs are computed by adding a normalization term to the logits (and hence **unaffected by temperature**);
- (2) Logprobs are logarithms of probabilities in sampling (and hence **predictably scaled with temperature**);
- (3) Same as in (2), **except for `temperature=0`**, when they are computed as in (1).

We observe (1) and (3) on OpenRouter-compatible providers; see code and example commands below.

It is worth examining whether logprobs interact with other sampling parameters in a similar fashion.

### Logprobs do not vary with temperature
```
$ uv run mwe.py --model meta-llama/llama-3.1-405b-instruct --runs 1 --provider lambda

Model: meta-llama/llama-3.1-405b-instruct
Prompt: The best ice cream flavor is (respond in a single word, DO NOT include any other text)
Providers: ['lambda']
================================================================================

Temperature: 0.0
----------------------------------------
  Run 1: Provider=Lambda | Token='Cookies' | logprob=-0.1823 | prob=83.34%
Average logprob: -0.1823 (probability: 83.34%)

Temperature: 0.0001
----------------------------------------
  Run 1: Provider=Lambda | Token='Cookies' | logprob=-0.1823 | prob=83.34%
Average logprob: -0.1823 (probability: 83.34%)

Temperature: 0.02
----------------------------------------
  Run 1: Provider=Lambda | Token='Cookies' | logprob=-0.1823 | prob=83.34%
Average logprob: -0.1823 (probability: 83.34%)

Temperature: 0.05
----------------------------------------
  Run 1: Provider=Lambda | Token='Cookies' | logprob=-0.1823 | prob=83.34%
Average logprob: -0.1823 (probability: 83.34%)

Temperature: 0.08
----------------------------------------
  Run 1: Provider=Lambda | Token='Cookies' | logprob=-0.1823 | prob=83.34%
Average logprob: -0.1823 (probability: 83.34%)

Temperature: 0.12
----------------------------------------
  Run 1: Provider=Lambda | Token='Cookies' | logprob=-0.1823 | prob=83.34%
Average logprob: -0.1823 (probability: 83.34%)
```

###  Logprobs scaled by temperature, except when temperature is 0
```bash
$ uv run mwe.py --model 01-ai/yi-large --runs 1 --provider fireworks

Model: 01-ai/yi-large
Prompt: The best ice cream flavor is (respond in a single word, DO NOT include any other text)
Providers: ['fireworks']
================================================================================

Temperature: 0.0
----------------------------------------
  Run 1: Provider=Fireworks | Token='Ch' | logprob=-0.7057 | prob=49.38%
Average logprob: -0.7057 (probability: 49.38%)

Temperature: 0.0001
----------------------------------------
  Run 1: Provider=Fireworks | Token='Ch' | logprob=0.0000 | prob=100.00%
Average logprob: 0.0000 (probability: 100.00%)

Temperature: 0.02
----------------------------------------
  Run 1: Provider=Fireworks | Token='Ch' | logprob=-0.0000 | prob=100.00%
Average logprob: -0.0000 (probability: 100.00%)

Temperature: 0.05
----------------------------------------
  Run 1: Provider=Fireworks | Token='Ch' | logprob=-0.0036 | prob=99.64%
Average logprob: -0.0036 (probability: 99.64%)

Temperature: 0.08
----------------------------------------
  Run 1: Provider=Fireworks | Token='Ch' | logprob=-0.0293 | prob=97.11%
Average logprob: -0.0293 (probability: 97.11%)

Temperature: 0.12
----------------------------------------
  Run 1: Provider=Fireworks | Token='Ch' | logprob=-0.0916 | prob=91.24%
Average logprob: -0.0916 (probability: 91.24%)
```

Together AI does something similar, but the scaling due to temperature is different. OpenRouter logprobs are not compatible with TogetherAI logprobs, and this issue is way too long already to include another code example, but can discuss further if this is relevant.

### Logprobs vary due to various reasons
OpenAI logprobs are highly nondeterministic. We include this behavior for completeness; it probably does not make sense to address it in Standard Completions, as this behavior cannot be compliant to any sort of standard on logprobs.

```
$ uv run mwe.py --model openai/gpt-4o-2024-11-20 --runs 3 --provider openai          

Model: openai/gpt-4o-2024-11-20
Prompt: The best ice cream flavor is (respond in a single word, DO NOT include any other text)
Providers: ['openai']
================================================================================

Temperature: 0.0
----------------------------------------
  Run 1: Provider=OpenAI | Token='Chocolate' | logprob=-0.4959 | prob=60.90%
  Run 2: Provider=OpenAI | Token='Chocolate' | logprob=-0.4959 | prob=60.90%
  Run 3: Provider=OpenAI | Token='Chocolate' | logprob=-0.4696 | prob=62.52%
Average logprob: -0.4871 (probability: 61.44%)

Temperature: 0.0001
----------------------------------------
  Run 1: Provider=OpenAI | Token='Chocolate' | logprob=-0.3691 | prob=69.13%
  Run 2: Provider=OpenAI | Token='Chocolate' | logprob=-0.3691 | prob=69.13%
  Run 3: Provider=OpenAI | Token='Chocolate' | logprob=-0.1466 | prob=86.36%
Average logprob: -0.2950 (probability: 74.46%)

Temperature: 0.02
----------------------------------------
  Run 1: Provider=OpenAI | Token='Chocolate' | logprob=-0.4696 | prob=62.52%
  Run 2: Provider=OpenAI | Token='Chocolate' | logprob=-0.4696 | prob=62.52%
  Run 3: Provider=OpenAI | Token='Chocolate' | logprob=-0.3691 | prob=69.13%
Average logprob: -0.4361 (probability: 64.65%)

Temperature: 0.05
----------------------------------------
  Run 1: Provider=OpenAI | Token='Chocolate' | logprob=-0.1557 | prob=85.58%
  Run 2: Provider=OpenAI | Token='Chocolate' | logprob=-0.4959 | prob=60.90%
  Run 3: Provider=OpenAI | Token='Chocolate' | logprob=-0.4696 | prob=62.52%
Average logprob: -0.3737 (probability: 68.82%)

Temperature: 0.08
----------------------------------------
  Run 1: Provider=OpenAI | Token='Chocolate' | logprob=-0.4696 | prob=62.52%
  Run 2: Provider=OpenAI | Token='Chocolate' | logprob=-0.4959 | prob=60.90%
  Run 3: Provider=OpenAI | Token='Chocolate' | logprob=-0.4959 | prob=60.90%
Average logprob: -0.4871 (probability: 61.44%)

Temperature: 0.12
----------------------------------------
  Run 1: Provider=OpenAI | Token='Chocolate' | logprob=-0.4696 | prob=62.52%
  Run 2: Provider=OpenAI | Token='Chocolate' | logprob=-0.1466 | prob=86.36%
  Run 3: Provider=OpenAI | Token='Chocolate' | logprob=-0.5385 | prob=58.36%
Average logprob: -0.3849 (probability: 68.05%)
```

## mwe.py
```python
#!/usr/bin/env uv run
# /// script
# dependencies = [
#   "requests",
#   "python-dotenv",
#   "simple-parsing",
# ]
# ///
"""
Minimal script for demonstrating how temperature affects logprobs.
"""

import os
import requests
import json
import math
from dataclasses import dataclass
from simple_parsing import ArgumentParser
from dotenv import load_dotenv

load_dotenv()


def call_model(model, prompt, temperature=0.0, providers=None):
    """Call OpenRouter API with specified temperature and providers."""
    api_key = os.getenv("OPENROUTER_API_KEY")

    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}

    data = {
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "logprobs": True,
        "top_logprobs": 5,
        "max_tokens": 1,
        "temperature": temperature,
    }

    if providers:
        data["provider"] = {"only": providers}

    response = requests.post(
        "https://openrouter.ai/api/v1/chat/completions", headers=headers, json=data
    )

    return response.json()


@dataclass(kw_only=True)
class Args:
    model: str = "01-ai/yi-large"
    prompt: str = "The best ice cream flavor is (respond in a single word, DO NOT include any other text)"
    runs: int = 5
    verbose: bool = False
    providers: list[str] | None = None


def main():
    parser = ArgumentParser()
    parser.add_arguments(Args, dest="args")
    args = parser.parse_args().args

    temperatures = [0.0, 0.0001, 0.02, 0.05, 0.08, 0.12]

    print(f"Model: {args.model}")
    print(f"Prompt: {args.prompt}")
    if args.providers:
        print(f"Providers: {args.providers}")
    print("=" * 80)

    for temp in temperatures:
        print(f"\nTemperature: {temp}\n" + "-" * 40)

        logprobs = []
        provider = None

        for run in range(args.runs):
            result = call_model(args.model, args.prompt, temp, providers=args.providers)

            if args.verbose:
                print(f"\n  Run {run + 1} - Full API Response:\n{json.dumps(result, indent=2)}")

            if "choices" in result and result["choices"]:
                choice = result["choices"][0]
                provider = result.get("provider", "Unknown")

                if choice.get("logprobs") and choice["logprobs"].get("content"):
                    token = choice["logprobs"]["content"][0]
                    logprob = token["logprob"]
                    logprobs.append(logprob)
                    prob = math.exp(logprob) * 100
                    print(
                        f"  Run {run + 1}: Provider={provider} | Token='{token['token']}' | logprob={logprob:.4f} | prob={prob:.2f}%"
                    )

        if logprobs:
            avg = sum(logprobs) / len(logprobs)
            print(f"Average logprob: {avg:.4f} (probability: {math.exp(avg) * 100:.2f}%)")


if __name__ == "__main__":
    main()
```
